BDArmory/Weapons/Missiles/MultiMissileLauncher.cs

```diff
@@ -0,0 +1,374 @@
+                                                    break;
+                                                }
+                                            }
+                                    }
+                                }
+                                TargetID++;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        if (wpm != null) wpm.SendTargetDataToMissile(ml);
+                    }
+                }
+                else
+                {
+                    wpm.SendTargetDataToMissile(ml);
+                }
+                ml.launched = true;
+                ml.TargetPosition = vessel.ReferenceTransform.position + (vessel.ReferenceTransform.up * 5000); //set initial target position so if no target update, missileBase will count a miss if it nears this point or is flying post-thrust
+                ml.MissileLaunch();
+                launchTransforms[m].localScale = Vector3.zero;
+                tubesFired++;
+            }
+            if (deployState != null)
+            {
+                deployState.enabled = true;
+                deployState.speed = -1;
+                yield return new WaitWhileFixed(() => deployState.normalizedTime > 0);
+                deployState.normalizedTime = 0;
+                deployState.speed = 0;
+                deployState.enabled = false;
+            }
+            if (tubesFired >= launchTransforms.Length) //add a timer for reloading a partially emptied MML if it hasn't been used for a while?
+            {
+                if (BDArmorySettings.INFINITE_AMMO || missileSpawner.ammoCount >= salvoSize)
+                    if (!(missileLauncher.reloadRoutine != null))
+                    {
+                        missileLauncher.reloadRoutine = StartCoroutine(missileLauncher.MissileReload());
+                        tubesFired = 0;
+                        if (BDArmorySettings.DEBUG_MISSILES) Debug.Log("[BDArmory.MultiMissileLauncher] all submunitions fired. Reloading");
+                    }
+            }
+            if (LaunchThenDestroy)
+            {
+                if (part != null)
+                {
+                    missileLauncher.DestroyMissile();
+                }
+            }
+            missileSalvo = null;
+        }
+
+        public void SetupMissileDummyPool(string modelpath)
+        {
+            if (mslDummyPool == null)
+                mslDummyPool = MissileDummy.CreateDummyPool(modelpath);
+        }
+    }
+}
+
+//huh. Vessel is a Monobehavior... Could a Vessel component be added to, say, rockets? if So, going the rocket route for missile pods/reloads would much simpler as most of the targeting code would no longer need to be rewritted to look for TargetInfos instead of Vessels. Would still need to mod the Bullet/Rocket/laser/Explosion hit code, and add an internal HP value, but...
```

Comment:
Maybe, but KSP does a lot behind the scenes for newly spawned vessels. I suspect that there'd be a significant performance hit in spawning them and some of this occurs over multiple frames (this is where the massive delay in setting up asteroid fields comes from until they're pooled and stripped of their excess modules).
Also, since KSP holds onto vessel and part references, this'd definitely increase the memory leak rate.

Reply:
Yeah, that's the one thing I'm worried about with the missile refactor, and why I was wondering about rockets - Initially I had thought about basically turning missiles into fancier versions of rockets - a pooled, non-part/vessel projectile that wouldn't have much performance impact, but would require a rewrite of most of the targeting/detection/missileUtilities code now that there's a class of target that doesn't have a 'Vessel' component and/or missiles now only targetable by anti-Missile APS, which adding a Vessel component to would avert at least in part.
As it stands, spawning in a single missile part/vessel now and again should be far less leaky than 60-100 part vessel*6 every heat start, but if some competition decided to go full Ace Combat silliness with dozens of heats featuring planes that are launching scores of missiles each...

Reply:
I'll have to run the missile spawning through the profiler at some point to see how much of an effect it's having on GC allocation and performance.
I wonder if it's possible to avoid a lot of the spawning cost and memory leakage by detaching the Vessel component from the missile prior to getting destroyed, packing and disabling it, then attaching, unpacking and enabling it on a new missile? Something to look at in the future...